pullman/storageproviders/gcs/downloader.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2021 IBM Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package gcsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-logr/logr"
+	"github.com/kserve/modelmesh-runtime-adapter/pullman"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type gcsClientFactory struct{}
+
+// gcsClientFactory implements gcsDownloaderFactory
+var _ gcsDownloaderFactory = (*gcsClientFactory)(nil)
+
+func (f gcsClientFactory) newDownloader(log logr.Logger, credentials map[string]string) (gcsDownloader, error) {
+	ctx := context.Background()
+	var cl *storage.Client
+	var err error
+
+	if len(credentials) > 0 {
+		credJson, marshalErr := json.Marshal(credentials)
+		if marshalErr != nil {
+			return nil, fmt.Errorf("unable to marshal json credentials: %w", marshalErr)
+		}
+		cl, err = storage.NewClient(ctx, option.WithCredentialsJSON(credJson))
+	} else {
+		cl, err = storage.NewClient(ctx, option.WithoutAuthentication())
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcsImplDownloader{
+		client: cl,
+		log:    log,
+	}, nil
+}
+
+type gcsImplDownloader struct {
+	client *storage.Client
+	log    logr.Logger
+
+	wg  sync.WaitGroup
+	mu  sync.Mutex
+	err error
+}
+
+func (d *gcsImplDownloader) listObjects(ctx context.Context, bucket string, prefix string) ([]string, error) {
+	query := &storage.Query{Prefix: prefix}
+	it := d.client.Bucket(bucket).Objects(ctx, query)
+
+	objectPaths := make([]string, 0, 10)
+
+	for {
+		obj, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("GCS listObjects: unable to list bucket %q: %v", bucket, err)
+		}
+
+		if !d.shouldIgnoreObject(obj, prefix) {
+			objectPaths = append(objectPaths, obj.Name)
+		}
+	}
+	return objectPaths, nil
+
+}
+
+func (d *gcsImplDownloader) downloadBatch(ctx context.Context, bucket string, targets []pullman.Target) error {
+	ch := make(chan pullman.Target, len(targets))
+	for _, target := range targets {
+		ch <- target
+	}
+
```

Comment:
Instead of using `downloadConcurrency` for the count below, better to use
```go
workers := math.Min(downloadConcurrency, len(targets))
```

Reply:
Apparently the  `math` function only has a `Min` function for `float64` and not integers. I just added a simple check to perform the same function instead of doing a bunch of type conversions.